cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/sidecar/GcpMetadataHttpHandlerInternal.java

```diff
@@ -139,31 +157,34 @@ public void token(HttpRequest request, HttpResponder responder,
       // needed when initializing
       // io.cdap.cdap.common.guice.DFSLocationModule$LocationFactoryProvider#get
       // in io.cdap.cdap.internal.app.worker.TaskWorkerTwillRunnable.
+      LOG.warn("The GCP Metadata Task Context has been identified as null.");
       GcpTokenResponse gcpTokenResponse = new GcpTokenResponse("Bearer", "invalidToken", 3599);
       responder.sendJson(HttpResponseStatus.OK, GSON.toJson(gcpTokenResponse));
       return;
     }
 
     try {
       // fetch token from credential provider
-      GcpTokenResponse gcpTokenResponse =
-          Retries.callWithRetries(() -> fetchTokenFromCredentialProvider(scopes),
-              RetryStrategies.fromConfiguration(cConf, Constants.Service.TASK_WORKER + "."));
+      ProvisionedCredential provisionedCredential =
+          credentialLoadingCache.get(
+              new ProvisionedCredentialCacheKey(this.gcpMetadataTaskContext, scopes));
+      GcpTokenResponse gcpTokenResponse = new GcpTokenResponse("Bearer",
+          provisionedCredential.get(),
+          Duration.between(Instant.now(), provisionedCredential.getExpiration()).getSeconds());
       responder.sendJson(HttpResponseStatus.OK, GSON.toJson(gcpTokenResponse));
       return;
-    } catch (NotFoundException e) {
-      // if credential identity not found,
-      // fallback to gcp metadata server for backward compatibility.
-    } catch (Exception ex) {
-      if (ex instanceof HttpErrorStatusProvider) {
-        HttpResponseStatus status = HttpResponseStatus.valueOf(
-            ((HttpErrorStatusProvider) ex).getStatusCode());
-        responder.sendJson(status, exceptionToJson(ex));
+    } catch (ExecutionException e) {
+      if (e.getCause() instanceof NotFoundException) {
+        // if credential identity not found,
+        // fallback to gcp metadata server for backward compatibility.
+      } else if (e.getCause() instanceof HttpErrorStatusProvider) {
+        throw e;
       } else {
-        LOG.warn("Failed to fetch token from credential provider", ex);
-        responder.sendJson(HttpResponseStatus.INTERNAL_SERVER_ERROR, exceptionToJson(ex));
+        LOG.warn("Failed to fetch token from credential provider", e.getCause());
+        responder.sendJson(HttpResponseStatus.INTERNAL_SERVER_ERROR,
```

Comment:
I believe you can also just propagate this for the exception handler to handle it (and the logging too).

Reply:
ack.